### PR TITLE
Migration guide from Guard to MaestroHdfs

### PR DESCRIPTION
**Pull request by toddmowen**
Some docs for downstream users. ~~Waiting on finalisation of #370.~~

### CODE REVIEW

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/MaestroHdfs.scala [hunk 1]
```diff
@@ -23,7 +23,105 @@ import au.com.cba.omnia.permafrost.hdfs.Hdfs
 import Guard.{NotProcessed, IngestionComplete}
 
 /**
- * Utility functions for Hdfs monad (the au.com.cba.omnia.permafrost.hdfs.Hdfs), with a maestro flavour.
+ * Utility functions for Hdfs monad
```
Comment (sujeshce) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/MaestroHdfs.scala]: Nice documentation :+1:

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/Guard.scala [hunk 2]
```diff
@@ -25,6 +25,11 @@ case class GuardFilter(filter: (FileSystem, Path) => Boolean) {
     GuardFilter((fs, p) => filter(fs, p) && that.filter(fs, p))
 }
 
+/**
+ * (DEPRECATED) Utililty functions that operate on the Hadoop filesystem
```
Comment (vineethvarghese) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/Guard.scala]: Spelling mistake - Utililty

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/Guard.scala [hunk 3]
```diff
@@ -25,6 +25,11 @@ case class GuardFilter(filter: (FileSystem, Path) => Boolean) {
...
+ *
+ * <i>Use</i> [[MaestroHdfs]] <i>instead (same method names, but return type is Hdfs).</i>
```
Comment (vineethvarghese) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/Guard.scala]: Generally for scala doc we don't really use html tags https://wiki.scala-lang.org/display/SW/Syntax

File: maestro-example/src/main/scala/au/com/cba/omnia/maestro/example/TransformExecution.scala [hunk 4]
```diff
@@ -61,7 +61,7 @@ object TransformExecution {
     )
     for {
       conf                <- Execution.getConfig.map(TransformConfig(_))
-      files               <- Execution.from(Guard.expandPaths(conf.inputGlob))
+      files               <- Execution.fromHdfs(MaestroHdfs.expandPaths(conf.inputGlob))
```
Comment (vineethvarghese) [inline maestro-example/src/main/scala/au/com/cba/omnia/maestro/example/TransformExecution.scala]: Maybe use the short form `Execution.hdfs` method?

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/MaestroHdfs.scala [hunk 5]
```diff
@@ -23,7 +23,105 @@ import au.com.cba.omnia.permafrost.hdfs.Hdfs
...
+ * (<a href="https://commbank.github.io/permafrost/latest/api/index.html#au.com.cba.omnia.permafrost.hdfs.Hdfs">
```
Comment (vineethvarghese) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/MaestroHdfs.scala]: How about scala doc tags rather than html tags?

**Comment by vineethvarghese** [comment]
@toddmowen In general we use scala doc format tags rather than html format tags. Is the code sample in the documentation from a real internal etl project?

**Comment by stephanh** [comment]
@toddmowen what is the state of this PR? Are you still planning on merging it?

**Comment by toddmowen** [comment]
Yesterday I began fixing it up in response to @vineethvarghese's comments. Soon I want to merge it, but not in the current state.

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/MaestroHdfs.scala [hunk 6]
```diff
@@ -24,12 +24,91 @@ import au.com.cba.omnia.permafrost.hdfs.Hdfs
...
 
 /**
- * Utility functions for Hdfs monad (the au.com.cba.omnia.permafrost.hdfs.Hdfs), with a maestro flavour.
- */
+  * Utility functions for the [[https://commbank.github.io/permafrost/latest/api/index.html#au.com.cba.omnia.permafrost.hdfs.Hdfs Hdfs monad]],
+  * with a Maestro flavour.
+  *
+  * This object was introduced in Maestro 2.10 to replace the [[Guard]] object, which served a similar purpose but
+  * predated the Hdfs monad. [[Guard]] is now deprecated and ''will be removed'' in a future version. Users of
+  * Maestro should update any code which uses [[Guard]] to use [[MaestroHdfs]] instead.
```
Comment (toddmowen) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/hdfs/MaestroHdfs.scala]: (I actually have no idea if it predated the Hdfs monad, but it sounds plausible)

**Comment by vineethvarghese** [comment]
Look fine to me :+1: . Do you want to add a deprecated compiler warning for `Guard`?

**Comment by toddmowen** [comment]
@vineethvarghese I deliberately chose not to add `@deprecated` to the object itself, but added it in maestro-api instead (so ultimately users will still get the compiler warnings).

It's not really ideal. But if I annotate the object directly, then compiling maestro-api will emit the compiler warning.

**Comment by vineethvarghese** [comment]
Sounds good :)

**Comment by toddmowen** [comment]
Merged into `master` (target of this PR was a different branch, and cannot be changed).

**Pull request closed without merging**